Convert Netty IOExceptions to UNAVAILABLE.

nmittler opened the pull request:
Also inspecting the entire causal chain to identify the best possible Status.

Fixes #1053
Fixes #1522

nmittler commented:
@ejona86 @carl-mastrangelo PTAL

nmittler commented:
FYI this should also fix #1522

ejona86 commented on netty/src/main/java/io/grpc/netty/Utils.java:
```diff
@@ -168,26 +169,33 @@ private static Http2Headers convertMetadata(Metadata headers) {
+
+    // Iterate over the causal chain for the error and find the best status.
+    for (Throwable cause : Throwables.getCausalChain(t)) {
+      if (cause instanceof IOException) {
```
Please include the ClosedChannelException code. Yes, it is disabled, but there is another bug to try to find the stragglers that prevent making it UNKNOWN instead of UNAVAILABLE (since ClosedChannelException is an IOException).

nmittler replied:
done.

ejona86 commented on netty/src/main/java/io/grpc/netty/Utils.java:
```diff
@@ -168,26 +169,33 @@ private static Http2Headers convertMetadata(Metadata headers) {
-      return Status.INTERNAL.withCause(t);
+
+    // Assume UNKNOWN unless we find a better status.
+    String msg = t.getMessage();
```
Don't copy the msg to the status description.

carl-mastrangelo replied:
There needs to be some message, since the stack trace won't make it over the wire.

ejona86 replied:
These are local-delivery only. Server cancellations caused by this turn into RST_STREAMS which have no details. In addition, we _really_ shouldn't send arbitrary strings seen in exceptions to remote endpoints.

nmittler replied:
done.

ejona86 commented on netty/src/main/java/io/grpc/netty/Utils.java:
```diff
@@ -168,26 +169,33 @@ private static Http2Headers convertMetadata(Metadata headers) {
+      if (cause instanceof Http2Exception) {
+        // Use INTERNAL in case we can't find anything better.
+        msg = msg != null ? msg : cause.getMessage();
+        status = Status.INTERNAL.withCause(t).withDescription(msg);
```
I'm not sure how comfortable I am with this, since it favors the deepest exception. Are Http2Exceptions included as the cause of other Http2Exceptions?

ejona86 replied:
Oh, I see. It is still using `t` instead of `cause`. That works for me, since it is only changing the status code.

carl-mastrangelo replied:
There should be a return here too right?

ejona86 replied:
The problem is there _can't_ be a return here, as it would defeat the point. We're dealing with IOExceptions being wrapped in Http2Exceptions.

carl-mastrangelo replied:
Then why is it rebuilding the status each time through the loop?

P.S.  github comments obscure the heck out of this code.

ejona86 replied:
The rebuilding of the status each time is part of why I was uncomfortable with it. I'm still uncomfortable with it, but not as uncomfortable with it now that I see it uses `t` instead of `cause` as the Status's cause.

P.S. Uncheck "Show notes" at the top right of this file's diff (next to view, across from the file name).

carl-mastrangelo commented on netty/src/test/java/io/grpc/netty/UtilsTest.java:
```diff
@@ -43,20 +44,50 @@
 public class UtilsTest {
   @Test
-  public void testStatusFromThrowable() {
+  public void testStatusFromNestedStatusException() {
```
nit: junit4 tests shouldn't use "test" prefix.  They already have `@Test` above.

nmittler replied:
done.

carl-mastrangelo commented on netty/src/test/java/io/grpc/netty/UtilsTest.java:
```diff
@@ -43,20 +44,50 @@
     t = new ConnectTimeoutException("msg");
-    assertStatusEquals(Status.UNAVAILABLE.withCause(t), Utils.statusFromThrowable(t));
+    assertStatusEquals(Status.UNAVAILABLE.withCause(t).withDescription("msg"),
+            Utils.statusFromThrowable(t));
```
nit: indent -4

nmittler replied:
done.

carl-mastrangelo commented:
Some nits, but otherwise LGTM

nmittler commented:
@ejona86 @carl-mastrangelo PTAL. I've updated the following:

1. Removed use of guava's `getCausalChain` due to #1663. 
2. Removed the rebuilding of Status during each iteration. Now just saving the `Status.Code` and build once at the end.

carl-mastrangelo commented:
@nmittler still LGTM

nmittler commented:
@carl-mastrangelo @ejona86 based on discussion with @ejona86 I've moved the change to `AbstractNettyHandler`. The new code avoids the forced wrapping of exceptions with `Http2Exception` entirely.

ejona86 commented:
@nmittler LGTM

nmittler commented:
Cherry-picked as 22e719a
